earth_enterprise/src/common/khCache.h

```diff
@@ -162,24 +171,59 @@ class khCache {
  public:
   typedef typename MapType::size_type size_type;
   size_type size(void) const { return map.size(); }
-  size_type capacity(void) const { return targetMax; }
+  uint64 capacity(void) const { return targetMax; }
+  uint64 getMemoryUse(void) const { return cacheObjectSizes; }
+  uint64 itemSizeMinusKeyAndValue(void) const { return sizeof(khCacheItem<Key, Value>) - sizeof(Key) - sizeof(Value); }
 
-  khCache(uint targetMax_
+  khCache(uint64 targetMax_
 #ifdef SUPPORT_VERBOSE
           , bool verbose_ = false
 #endif
           ) : head(0), tail(0), targetMax(targetMax_),
 #ifdef SUPPORT_VERBOSE
               verbose(verbose_),
 #endif
-              numItems(0)
+              numItems(0), cacheObjectSizes(0),
+              limitCacheMemory(0),
+              khCacheItemSize(itemSizeMinusKeyAndValue())
 
   {
     CheckListInvariant();
   }
   ~khCache(void) {
     clear();
   }
+  template<class T>
+  uint64 getKeyOrValueSize(const T obj) {
+    return sizeof(obj);
+  }
+  uint64 getKeyOrValueSize(const std::string str) {
+    return (sizeof(str) + (str.capacity() * sizeof(char)));
+  }
+  uint64 getKeyOrValueSize(const SharedString shstr) {
+    return shstr.GetSharedStringSize();
+  }
+  template<class T>
+  uint64 getKeyOrValueSize(const khRefGuard<T> obj) {
+    return (sizeof(obj) + obj.getRefGuardSize());
+  }
+  void updateObjectSize(const Key &key) {
+    if ( FindItem(key) != 0 ) {
+      Item *item = FindItem(key);
+      uint64 size = khCacheItemSize + getKeyOrValueSize(item->key) + getKeyOrValueSize(item->val);
+      if (size > item->size) {
+        cacheObjectSizes += (size - item->size);
```

Comment:
I'm pretty sure that both of these cases simplify to `cacheObjectSizes = cacheObjectSizes - item->size + size`.